python/common_tool_methods.py

```diff
@@ -0,0 +1,146 @@
+import re
+import subprocess
+from datetime import datetime
+from typing import Dict, List
+import os
+
+from github import Repository, PullRequest
+from jinja2 import Environment, FileSystemLoader
+
+from .common_validations import *
+
+
+def get_spec_file_name(project_name: str) -> str:
+    return f"{project_name}.spec"
+
+
+def get_version_number(version: str, fancy: bool, fancy_release_count: int) -> str:
+    fancy_suffix = f"-{fancy_release_count}" if fancy else ""
+    return f"{version}{fancy_suffix}"
+
+
+def get_version_number_with_project_name(project_name: str, version: str, fancy: bool, fancy_release_count: int) -> str:
+    fancy_suffix = f"-{fancy_release_count}" if fancy else ""
+    return f"{version}.{project_name}{fancy_suffix}"
+
+
+def get_template_environment(template_dir: str) -> Environment:
+    file_loader = FileSystemLoader(template_dir)
+    env = Environment(loader=file_loader)
+    return env
+
+
+def find_nth_overlapping(subject_string, search_string, n) -> int:
+    start = subject_string.find(search_string)
+
+    while start >= 0 and n > 1:
+        start = subject_string.find(search_string, start + 1)
+        n -= 1
+    return start
+
+
+def find_nth_overlapping_line_by_regex(subject_string, regex_pattern, n) -> int:
+    lines = subject_string.splitlines()
+    counter = 0
+    for i in range(len(lines)):
+        if re.match(regex_pattern, lines[i]):
+            counter = counter + 1
+        if counter == n:
+            return i
+    return -1
```

Comment:
subject_string is str type I added str type
I didn't change the parameter name of find_nth_matching_line_number into lines because I think splitting into lines should be in this functions
Others are fixed

Reply:
I couldn't quite understand what is happening around `find_nth_matching_line_number` & `find_nth_matching_line`.
`find_nth_matching_line` splits the lines and calls `find_nth_matching_line_number`.
However, `find_nth_matching_line_number` splits those lines too. 
I wonder how our unit tests don't fail on that.

Also a minor comment: maybe return `None` instead of empty string.

In the end, maybe we can implement an helper function and make use of it:

```python
def find_nth_matching_line(subject_string: str, regex_pattern: str, n: int) -> str:
    """Takes a subject string, regex param and the search index as parameter and returns line content of found match.
        If not found returns None"""
    lines = subject_string.splitlines()
    line_number = find_nth_matching_string_idx(subject_string, regex_pattern, n)
    if line_number != -1:
        return lines[line_number]
    else:
        return None


def find_nth_matching_line_number(subject_string: str, regex_pattern: str, n: int) -> int:
    """Takes a subject string, regex param and the search index as parameter and returns line number of found match.
        If not found returns -1"""
    lines = subject_string.splitlines()
    line_number = find_nth_matching_string_and_idx(lines, regex_pattern, n)
    return line_number


def find_nth_matching_string_idx(string_list: str, regex_pattern: str, n: int):
    counter = 0
    for idx, string in enumerate(string_list):
        if re.match(regex_pattern, string):
            counter = counter + 1
        if counter == n:
            return idx
    return -1
```

Reply:
I unified method to make these methods less complex and named find_nth_matching_line_and_line_number. Is it acceptable?
I usually avoid Null return types since it is not suggested as clean code principles to avoid futher null checks. But if there is a team policy I will obey this

Reply:
I don't think we have a general rule for this. Each project we maintain have slightly different set of principles.

Personally I am fine with this piece of code either way.

Reply:
By the way "do not return Null" value is a rule in "Clean Code" book If I can find, I will share it with you